src/hmm/transition-model.cc

```diff
@@ -24,74 +24,122 @@
-    for (int32 j = 0; j < static_cast<int32>(entry.size()); j++) {  // for each state...
-      int32 pdf_class = entry[j].pdf_class;
-      if (pdf_class != kNoPdf) {
-        to_hmm_state_list[std::make_pair(phone, pdf_class)].push_back(j);
+
+  if (SelfLoopEqualsForward()) {
+    // this branch deals with when self_loop_pdf_class is always the same as the pdf_class
+    std::vector<std::vector<std::pair<int32, int32> > > pdf_info;
+    std::vector<int32> num_pdf_classes( 1 + *std::max_element(phones.begin(), phones.end()), -1);
+    for (size_t i = 0; i < phones.size(); i++)
+      num_pdf_classes[phones[i]] = topo_.NumPdfClasses(phones[i]);
+    ctx_dep.GetPdfInfo(phones, num_pdf_classes, &pdf_info);
+    // pdf_info is list indexed by pdf of which (phone, pdf_class) it
+    // can correspond to.
+
+    std::map<std::pair<int32, int32>, std::vector<int32> > to_hmm_state_list;
+    // to_hmm_state_list is a map from (phone, pdf_class) to the list
+    // of hmm-states in the HMM for that phone that that (phone, pdf-class)
+    // can correspond to.
+    for (size_t i = 0; i < phones.size(); i++) {  // setting up to_hmm_state_list.
+      int32 phone = phones[i];
+      const HmmTopology::TopologyEntry &entry = topo_.TopologyForPhone(phone);
+      for (int32 j = 0; j < static_cast<int32>(entry.size()); j++) {  // for each state...
+        int32 pdf_class = entry[j].pdf_class;
+        if (pdf_class != kNoPdf) {
+          to_hmm_state_list[std::make_pair(phone, pdf_class)].push_back(j);
+        }
       }
     }
-  }
-  for (int32 pdf = 0; pdf < static_cast<int32>(pdf_info.size()); pdf++) {
-    for (size_t j = 0; j < pdf_info[pdf].size(); j++) {
-      int32 phone = pdf_info[pdf][j].first,
-          pdf_class = pdf_info[pdf][j].second;
-      const std::vector<int32> &state_vec = to_hmm_state_list[std::make_pair(phone, pdf_class)];
-      KALDI_ASSERT(!state_vec.empty());
-      // state_vec is a list of the possible HMM-states that emit this
-      // pdf_class.
-      for (size_t k = 0; k < state_vec.size(); k++) {
-        int32 hmm_state = state_vec[k];
-        triples_.push_back(Triple(phone, hmm_state, pdf));
+
+    for (int32 pdf = 0; pdf < static_cast<int32>(pdf_info.size()); pdf++) {
+      for (size_t j = 0; j < pdf_info[pdf].size(); j++) {
+        int32 phone = pdf_info[pdf][j].first,
+            pdf_class = pdf_info[pdf][j].second;
+        const std::vector<int32> &state_vec = to_hmm_state_list[std::make_pair(phone, pdf_class)];
+        KALDI_ASSERT(!state_vec.empty());
+        // state_vec is a list of the possible HMM-states that emit this
+        // pdf_class.
+        for (size_t k = 0; k < state_vec.size(); k++) {
+          int32 hmm_state = state_vec[k];
+          tuples_.push_back(Tuple(phone, hmm_state, pdf, pdf));
+        }
       }
     }
-  }
+  } else {
+    std::vector<std::vector<std::vector<std::pair<int32, int32> > > > pdf_info;
+    std::vector<std::vector<std::pair<int32, int32> > > pdf_class_pairs;
```

Comment:
need a comment saying what pdf_info represents.-- and of course you have to figure out pdf_class_pairs, and also explain what it is.